clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/command/amqp/ProtonBasedCommandConsumerFactory.java

```diff
@@ -0,0 +1,247 @@
+import org.eclipse.hono.adapter.client.amqp.AbstractServiceClient;
+import org.eclipse.hono.adapter.client.command.Command;
+import org.eclipse.hono.adapter.client.command.CommandConsumer;
+import org.eclipse.hono.adapter.client.command.CommandConsumerFactory;
+import org.eclipse.hono.adapter.client.command.CommandContext;
+import org.eclipse.hono.client.CommandTargetMapper;
+import org.eclipse.hono.client.HonoConnection;
+import org.eclipse.hono.client.ProtocolAdapterCommandConsumerFactory;
+import org.eclipse.hono.client.ProtocolAdapterCommandConsumerFactory.CommandHandlingAdapterInfoAccess;
+import org.eclipse.hono.client.SendMessageSampler.Factory;
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.util.Constants;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonHelper;
+
+
+/**
+ * A vertx-proton based factory for creating consumers of command messages received via the
+ * AMQP 1.0 Messaging Network.
+ * <p>
+ * This implementation wraps a {@link ProtocolAdapterCommandConsumerFactory} and thus also supports
+ * routing of commands to a target protocol adapter instance.
+ *
+ */
+public class ProtonBasedCommandConsumerFactory extends AbstractServiceClient implements CommandConsumerFactory {
+
+    private final ProtocolAdapterCommandConsumerFactory factory;
+
+    /**
+     * Creates a new client for a connection.
+     *
+     * @param connection The connection to the AMQP 1.0 Messaging Network.
+     * @param samplerFactory The factory for creating samplers for tracing AMQP messages being sent.
+     * @param adapterConfig The protocol adapter's configuration properties.
+     * @param commandTargetMapper The component for mapping an incoming command to the gateway (if applicable) and
+     *            protocol adapter instance that can handle it.
+     * @param commandRoutingInfoAccess The component for setting and clearing information that maps a device to
+     *                                 a protocol adapter instance.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public ProtonBasedCommandConsumerFactory(
+            final HonoConnection connection,
+            final Factory samplerFactory,
+            final ProtocolAdapterProperties adapterConfig,
+            final CommandTargetMapper commandTargetMapper,
+            final CommandHandlingAdapterInfoAccess commandRoutingInfoAccess) {
+
+        super(connection, samplerFactory, adapterConfig);
+
+        Objects.requireNonNull(commandTargetMapper);
+        Objects.requireNonNull(commandRoutingInfoAccess);
+
+        factory = ProtocolAdapterCommandConsumerFactory.create(connection, samplerFactory);
+        factory.initialize(commandTargetMapper, commandRoutingInfoAccess);
+    }
+
+    private static class CommandContextAdapter implements CommandContext {
```

Comment:
IMHO it would help readability to move this inner class to the bottom of the class.